File: Medium/1846. Maximum Element After Decreasing and Rearranging/1846. Maximum Element After Decreasing and Rearranging.ts
function maximumElementAfterDecrementingAndRearranging(arr: number[]): number {
    arr.sort((a, b) => a - b); // Step 1

    let prev = 0;
    for (let i = 0; i < arr.length; i++) {
        arr[i] = Math.min(prev + 1, arr[i]); // Step 2
        prev = arr[i];
    }

    return Math.max(...arr); // Step 3
}

File: Medium/1846. Maximum Element After Decreasing and Rearranging/1846. Maximum Element After Decreasing and Rearranging.java
import java.util.Arrays;

public class Solution {
    public int maximumElementAfterDecrementingAndRearranging(int[] arr) {
        Arrays.sort(arr);  // Step 1

        int prev = 0;
        for (int i = 0; i < arr.length; i++) {
            arr[i] = Math.min(prev + 1, arr[i]);  // Step 2
            prev = arr[i];
        }

        int max = 0;
        for (int num : arr) {
            max = Math.max(max, num);
        }

        return max;  // Step 3
    }
}
